rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/wait.hpp

```diff
@@ -15,15 +15,103 @@
 
 #include "rmw/error_handling.h"
 #include "rmw/impl/cpp/macros.hpp"
 #include "rmw/types.h"
 
 #include "rmw_connext_shared_cpp/condition_error.hpp"
+#include "rmw_connext_shared_cpp/event_converter.hpp"
 #include "rmw_connext_shared_cpp/types.hpp"
 #include "rmw_connext_shared_cpp/visibility_control.h"
+#include "rmw_connext_shared_cpp/connext_static_event_info.hpp"
+
+
+rmw_ret_t __gather_event_conditions(
+  rmw_events_t * events,
+  std::unordered_set<DDS::StatusCondition *> & status_conditions)
+{
+  RMW_CHECK_ARGUMENT_FOR_NULL(events, RMW_RET_INVALID_ARGUMENT);
+
+  std::unordered_map<DDS::StatusCondition *, DDS::StatusMask> status_mask_map;
+  // gather all status conditions and masks
+  for (size_t i = 0; i < events->event_count; ++i) {
+    rmw_event_t * current_event = static_cast<rmw_event_t *>(events->events[i]);
+    DDS::Entity * dds_entity = static_cast<ConnextCustomEventInfo *>(
+      current_event->data)->get_entity();
+    if (!dds_entity) {
+      RMW_SET_ERROR_MSG("Event handle is null");
+      return RMW_RET_ERROR;
+    }
+    DDS::StatusCondition * status_condition = dds_entity->get_statuscondition();
+    if (!status_condition) {
+      RMW_SET_ERROR_MSG("status condition handle is null");
+      return RMW_RET_ERROR;
+    }
+    if (is_event_supported(current_event->event_type)) {
+      if (status_mask_map.find(status_condition) == status_mask_map.end()) {
+        status_mask_map[status_condition] = DDS::STATUS_MASK_NONE;
+      }
+      status_mask_map[status_condition] = status_mask_map[status_condition] |
+        get_status_kind_from_rmw(current_event->event_type);
+    }
+  }
+  for (auto & pair : status_mask_map) {
+    // set the status condition's mask with the supported type
+    pair.first->set_enabled_statuses(pair.second);
+    status_conditions.insert(pair.first);
+  }
+  return RMW_RET_OK;
+}
+
+rmw_ret_t __handle_active_event_conditions(rmw_events_t * events)
+{
+  // enable a status condition for each event
+  if (events) {
+    for (size_t i = 0; i < events->event_count; ++i) {
+      rmw_event_t * current_event = static_cast<rmw_event_t *>(events->events[i]);
+      DDS::Entity * dds_entity = static_cast<ConnextCustomEventInfo *>(
+        current_event->data)->get_entity();
+      if (!dds_entity) {
+        RMW_SET_ERROR_MSG("Event handle is null");
+        return RMW_RET_ERROR;
+      }
+
+      DDS::StatusMask status_mask = dds_entity->get_status_changes();
+      bool is_active = false;
+
+      if(is_event_supported(current_event->event_type)) {
+        is_active = static_cast<bool>(status_mask &
+          get_status_kind_from_rmw(current_event->event_type));
+      }
+      // if status condition is not found in the active set
+      // reset the subscriber handle
+      if (!is_active) {
+        events->events[i] = nullptr;
+      }
+    }
+  }
+  return RMW_RET_OK;
+}
+
+rmw_ret_t __detach_condition(
+  DDS::WaitSet * dds_wait_set,
+  DDSCondition * condition)
```

Comment:
Should this be `DDS::Condition`? I had suggested `DDS*` since I think that's mostly what was in the project before, but since you guys decided to go with `DDS::*` (which is fine) just try to be consistent locally (within functions and files).

Reply:
Sounds good, will change